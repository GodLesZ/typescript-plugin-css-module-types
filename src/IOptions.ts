import { SyncOptions as NodeSassOptions } from "node-sass";
import { ILoggerOptions } from "./Logger";

export default interface IOptions {
    camelCase?: CamelCaseOptions;
    customMatcher?: string;
    log?: ILoggerOptions;
    sass?: {
        renderOptions?: Partial<NodeSassOptions>;
    };
}

export type CamelCaseOptions = | true | "dashes" | "dashesOnly" | "only" | undefined;
