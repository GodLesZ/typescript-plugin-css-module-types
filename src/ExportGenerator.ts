import { CSSExports } from "icss-utils";
import { camelCase } from "lodash";
import reserved from "reserved-words";
import IOptions from "./IOptions";

const dashesCamelCase = (className: string) => {
    return className.replace(/-+(\w)/g, (match, firstLetter) => firstLetter.toUpperCase());
};

export default class ExportGenerator {
    private static readonly NOT_CAMELCASE_REGEXP = /[\-_]/;

    private readonly options: IOptions;

    public constructor(options: IOptions) {
        this.options = options;
    }

    public generate(classes: CSSExports) {
        const isCamelCase = (className: string) => !ExportGenerator.NOT_CAMELCASE_REGEXP.test(className);
        const isReservedWord = (className: string) => !reserved.check(className);

        const classNames = Object.keys(classes)
            .map(className => this.transformClassName(className))
            // Flatter array because transformClassName() may return multiple classes
            .reduce((previousValue: Array<string> = [], currentValue: Array<string>) => {
                return previousValue.concat(currentValue);
            }, []);

        const camelCasedKeys = classNames
            .filter(isCamelCase)
            .filter(isReservedWord)
            .map(className => `export const ${className}: string;`);

        const defaultExport = `\
declare const styles: {
    ${classNames.map(className => `"${className}": string;`).join("\n  ")}
};
export default styles;
`;

        if (camelCasedKeys.length) {
            return `${defaultExport}${camelCasedKeys.join("\n")}\n`;
        }

        return defaultExport;
    }

    public transformClassName(className: string) {
        const entries: Array<string> = [];

        switch (this.options.camelCase) {
            case true:
                entries.push(className);
                const camelCaseClassName = camelCase(className);
                if (camelCaseClassName !== className) {
                    entries.push(camelCaseClassName);
                }
                break;

            case "dashes":
                entries.push(className);
                const dashesCamelCaseClassName = dashesCamelCase(className);
                if (dashesCamelCaseClassName !== className) {
                    entries.push(dashesCamelCaseClassName);
                }
                break;

            case "only":
                entries.push(camelCase(className));
                break;

            case "dashesOnly":
                entries.push(dashesCamelCase(className));
                break;

            default:
                entries.push(className);
                break;
        }

        return entries;
    }
}
