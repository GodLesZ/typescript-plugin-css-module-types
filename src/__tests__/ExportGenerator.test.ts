import ExportGenerator from "../ExportGenerator";
import { CamelCaseOptions } from "../IOptions";

describe("ExportGenerator", () => {
    const classNames = [
        "class-name-a",
        "classNameB",
        "class-Name-C",
        "__class_nAmeD--",
    ];
    const tests: Array<CamelCaseOptions> = [true, "dashes", "dashesOnly", "only"];

    it(`should not transform classes when no option is set`, () => {
        const generator = new ExportGenerator({});
        const transformedClasses = classNames.map(className => generator.transformClassName(className));

        expect(transformedClasses).toMatchSnapshot();
    });

    tests.forEach(option => {
        it(`should transform classes correctly when \`camelCase\` set to \`${option}\``, () => {
            const generator = new ExportGenerator({
                camelCase: option,
            });
            const transformedClasses = classNames.map(className => generator.transformClassName(className));

            expect(transformedClasses).toMatchSnapshot();
        });
    });

    it("creates exports without constants if no camel-case class was found", () => {
        const generator = new ExportGenerator({});
        const exports = generator.generate({
            "some-class": "some-class----HASH",
            "some-class-2": "some-class-2----HASH",
            "const": "const----HASH",
        });

        expect(exports).toMatchSnapshot();
    });

    it("creates exports with constants for every camel-case class", () => {
        const generator = new ExportGenerator({});
        const exports = generator.generate({
            "camelCase": "camel-case----HASH",
            "camelCase2": "camel-case-2----HASH",
            "some-class": "some-class----HASH",
            "const": "const----HASH",
        });

        expect(exports).toMatchSnapshot();
    });
});
