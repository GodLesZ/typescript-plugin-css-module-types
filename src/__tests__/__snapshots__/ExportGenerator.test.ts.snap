// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportGenerator creates exports correctly 1`] = `
"declare const styles: {
    \\"camelCase\\": string;
  \\"camelCase2\\": string;
  \\"some-class\\": string;
  \\"const\\": string;
};
export default styles;
export const camelCase: string;
export const camelCase2: string;
"
`;

exports[`ExportGenerator creates exports with constants for every camel-case class 1`] = `
"declare const styles: {
    \\"camelCase\\": string;
  \\"camelCase2\\": string;
  \\"some-class\\": string;
  \\"const\\": string;
};
export default styles;
export const camelCase: string;
export const camelCase2: string;
"
`;

exports[`ExportGenerator creates exports without constants if no camel-case class was found 1`] = `
"declare const styles: {
    \\"some-class\\": string;
  \\"some-class-2\\": string;
  \\"const\\": string;
};
export default styles;
"
`;

exports[`ExportGenerator should not transform classes when no option is set 1`] = `
Array [
  Array [
    "class-name-a",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "class-Name-C",
  ],
  Array [
    "__class_nAmeD--",
  ],
]
`;

exports[`ExportGenerator should transform classes correctly when \`camelCase\` set to \`dashes\` 1`] = `
Array [
  Array [
    "class-name-a",
    "classNameA",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "class-Name-C",
    "classNameC",
  ],
  Array [
    "__class_nAmeD--",
  ],
]
`;

exports[`ExportGenerator should transform classes correctly when \`camelCase\` set to \`dashesOnly\` 1`] = `
Array [
  Array [
    "classNameA",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "classNameC",
  ],
  Array [
    "__class_nAmeD--",
  ],
]
`;

exports[`ExportGenerator should transform classes correctly when \`camelCase\` set to \`only\` 1`] = `
Array [
  Array [
    "classNameA",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "classNameC",
  ],
  Array [
    "classNAmeD",
  ],
]
`;

exports[`ExportGenerator should transform classes correctly when \`camelCase\` set to \`true\` 1`] = `
Array [
  Array [
    "class-name-a",
    "classNameA",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "class-Name-C",
    "classNameC",
  ],
  Array [
    "__class_nAmeD--",
    "classNAmeD",
  ],
]
`;
